src/Neo/SmartContract/Native/Notary.cs

```diff
@@ -0,0 +1,331 @@
+// Copyright (C) 2015-2025 The Neo Project.
+//
+// Notary.cs file belongs to the neo project and is free
+// software distributed under the MIT software license, see the
+// accompanying file LICENSE in the main directory of the
+// repository or http://www.opensource.org/licenses/mit-license.php
+// for more details.
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+#pragma warning disable IDE0051
+
+using Neo.Cryptography;
+using Neo.Cryptography.ECC;
+using Neo.Extensions;
+using Neo.Network.P2P;
+using Neo.Network.P2P.Payloads;
+using Neo.Persistence;
+using Neo.SmartContract.Manifest;
+using Neo.VM;
+using Neo.VM.Types;
+using System;
+using System.Linq;
+using System.Numerics;
+using Array = Neo.VM.Types.Array;
+
+namespace Neo.SmartContract.Native
+{
+    /// <summary>
+    /// The Notary native contract used for multisignature transactions forming assistance.
+    /// </summary>
+    public sealed class Notary : NativeContract
+    {
+        /// <summary>
+        /// A default value for maximum allowed NotValidBeforeDelta. It is set to be
+        /// 20 rounds for 7 validators, a little more than half an hour for 15-seconds blocks.
+        /// </summary>
+        private const int DefaultMaxNotValidBeforeDelta = 140;
+        /// <summary>
+        /// A default value for deposit lock period.
+        /// </summary>
+        private const int DefaultDepositDeltaTill = 5760;
+        private const byte Prefix_Deposit = 1;
+        private const byte Prefix_MaxNotValidBeforeDelta = 10;
+
+        internal Notary() : base() { }
+
+        public override Hardfork? ActiveIn => Hardfork.HF_Echidna;
+
+        internal override ContractTask InitializeAsync(ApplicationEngine engine, Hardfork? hardfork)
+        {
+            if (hardfork == ActiveIn)
+            {
+                engine.SnapshotCache.Add(CreateStorageKey(Prefix_MaxNotValidBeforeDelta), new StorageItem(DefaultMaxNotValidBeforeDelta));
+            }
+            return ContractTask.CompletedTask;
+        }
+
+        internal override async ContractTask OnPersistAsync(ApplicationEngine engine)
+        {
+            long nFees = 0;
+            ECPoint[] notaries = null;
+            foreach (Transaction tx in engine.PersistingBlock.Transactions)
```

Comment:
We can check if `PersistingBlock` is null, neo express debugger fail with something similar in the past.